frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
+
+    int requestNumber = pendingTaskNumber * am.getConfiguration().getLauncherConfig().getAmSearchNodeBufferFactor();
+    List<Node> candidateNodes = new ArrayList<Node>();
+    SelectionResult result = new SelectionResult();
+    for (String nodeName : filteredNodes) {
+      candidateNodes.add(allNodes.get(nodeName));
+    }
+    Collections.sort(candidateNodes);
+    for (int i = 0; i < requestNumber && i < candidateNodes.size(); i++) {
+      Node select = candidateNodes.get(i);
+      Long gpuAttribute = requestResource.getGpuAttribute();
+      if (gpuAttribute == 0) {
+        gpuAttribute = selectCandidateGpuAttribute(select, requestResource.getGpuNumber());
+      }
+      result.addSelection(select.getHost(), gpuAttribute, select.getAvailableResource().getPortRanges());
     }
+    return result;
+  }
 
-    if (selectionResult != null) {
-      LOGGER.logInfo(
-          "select: Found a SelectionResult satisfies the Request: SelectionResult: [%s]",
-          selectionResult);
-    } else {
-      LOGGER.logWarning(
-          "select: Cannot found a SelectionResult satisfies the Request: " +
-              "Check whether the Request can be relaxed to RM");
-      String notRelaxLogPrefix = "select: The Request cannot be relaxed to RM: Reason: ";
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        throw new NotAvailableException(
-            String.format(notRelaxLogPrefix +
-                    "NodeGpuType is specified: Request NodeGpuType: [%s]",
-                requestNodeGpuType));
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String taskRoleName) throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: TaskRole: [%s] Resource: [%s]", taskRoleName, requestResource);
+    String requestNodeLabel = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
+    String requestNodeGpuType = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeGpuType();
+    int pendingTaskNumber = am.getStatusManager().getUnAssociatedTaskCount(taskRoleName);
+    List<ValueRange> reUsePorts = null;
+
+    // Prefer to use previous successfully allocated ports. if no successfully ports, try to re-use the "Requesting" ports.
+    if (am.getRequestManager().getTaskRoles().get(taskRoleName).getUseTheSamePorts()) {
+      reUsePorts = am.getStatusManager().getLiveAssociatedContainerPorts(taskRoleName);
+      if (ValueRangeUtils.getValueNumber(reUsePorts) <= 0 && requestedPortsCache.containsKey(taskRoleName)) {
+        reUsePorts = requestedPortsCache.get(taskRoleName);
+        // the cache only guide the next task to use previous requesting port.
+        requestedPortsCache.remove(taskRoleName);
       }
+    }
+    SelectionResult result = select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, reUsePorts);
+    if (!requestedPortsCache.containsKey(taskRoleName) && pendingTaskNumber > 1)
```

Comment:
> > 1 [](start = 75, length = 4)

pendingTaskNumber > 0

Reply:
The "if"  need brackets ()

---
In reply to: [174714320](https://github.com/Microsoft/pai/pull/318#discussion_r174714320) [](ancestors = 174714320)